File: frontend/src/lib/api/workflow.ts
// =============================================
// 工作流 API 客戶端
// =============================================

import { apiClient } from './client'

// ==================== Types ====================

export type ScheduleFrequency = 'daily' | 'weekly' | 'monthly' | 'custom'
export type ScheduleStatus = 'active' | 'paused' | 'completed' | 'failed'
export type ExecutionStatus = 'pending' | 'running' | 'completed' | 'failed' | 'skipped'
export type ReportType = 'price_analysis' | 'competitor_report' | 'sales_summary' | 'inventory_alert' | 'custom'

export interface ScheduledReport {
  id: string
  name: string
  description: string | null
  report_type: ReportType
  report_config: Record<string, unknown> | null
  frequency: ScheduleFrequency
  schedule_time: string | null
  schedule_day: number | null
  cron_expression: string | null
  timezone: string
  delivery_channels: Record<string, unknown> | null
  status: ScheduleStatus
  last_run_at: string | null
  next_run_at: string | null
  run_count: number
  success_count: number
  failure_count: number
  consecutive_failures: number
  source_conversation_id: string | null
  created_by: string | null
  created_at: string
  updated_at: string
}

export interface ReportExecution {
  id: string
  schedule_id: string
  status: ExecutionStatus
  scheduled_at: string
  started_at: string | null
  completed_at: string | null
  duration_ms: number | null
  report_content: string | null
  report_data: Record<string, unknown> | null
  delivery_status: Record<string, unknown> | null
  error_message: string | null
  retry_count: number
  created_at: string
}

export interface ScheduleCreateRequest {
  name: string
  description?: string
  report_type?: ReportType
  report_config?: Record<string, unknown>
  frequency: ScheduleFrequency
  schedule_time?: string
  schedule_day?: number
  cron_expression?: string
  timezone?: string
  delivery_channels?: Record<string, unknown>
}

export interface ScheduleUpdateRequest {
  name?: string
  description?: string
  report_type?: ReportType
  report_config?: Record<string, unknown>
  frequency?: ScheduleFrequency
  schedule_time?: string
  schedule_day?: number
  cron_expression?: string
  timezone?: string
  delivery_channels?: Record<string, unknown>
  status?: ScheduleStatus
}

export interface ScheduleListResponse {
  items: ScheduledReport[]
  total: number
  limit: number
  offset: number
}

export interface ExecutionListResponse {
  items: ReportExecution[]
  total: number
}

export interface ScheduleActionResponse {
  success: boolean
  message: string
  schedule: ScheduledReport | null
}

export interface TriggerResponse {
  success: boolean
  message: string
  execution_id: string | null
  task_id: string | null
}

export interface NextRunsPreview {
  schedule_id: string
  next_runs: string[]
}

// ==================== API Functions ====================

/**
 * 列出排程
 */
export async function listSchedules(params?: {
  status?: ScheduleStatus
  conversation_id?: string
  limit?: number
  offset?: number
}): Promise<ScheduleListResponse> {
  const searchParams = new URLSearchParams()
  if (params?.status) searchParams.set('status', params.status)
  if (params?.conversation_id) searchParams.set('conversation_id', params.conversation_id)
  if (params?.limit) searchParams.set('limit', params.limit.toString())
  if (params?.offset) searchParams.set('offset', params.offset.toString())

  const query = searchParams.toString()
  const response = await apiClient.get(`/workflow/schedules${query ? `?${query}` : ''}`)
  return response as unknown as ScheduleListResponse
}

/**
 * 獲取排程詳情
 */
export async function getSchedule(scheduleId: string): Promise<ScheduledReport> {
  const response = await apiClient.get(`/workflow/schedules/${scheduleId}`)
  return response as unknown as ScheduledReport
}

/**
 * 創建排程
 */
export async function createSchedule(data: ScheduleCreateRequest): Promise<ScheduledReport> {
  const response = await apiClient.post('/workflow/schedules', data)
  return response as unknown as ScheduledReport
}

/**
 * 更新排程
 */
export async function updateSchedule(
  scheduleId: string,
  data: ScheduleUpdateRequest
): Promise<ScheduledReport> {
  const response = await apiClient.patch(`/workflow/schedules/${scheduleId}`, data)
  return response as unknown as ScheduledReport
}

/**
 * 刪除排程
 */
export async function deleteSchedule(scheduleId: string): Promise<void> {
  await apiClient.delete(`/workflow/schedules/${scheduleId}`)
}

/**
 * 暫停排程
 */
export async function pauseSchedule(scheduleId: string): Promise<ScheduleActionResponse> {
  const response = await apiClient.post(`/workflow/schedules/${scheduleId}/pause`)
  return response as unknown as ScheduleActionResponse
}

/**
 * 恢復排程
 */
export async function resumeSchedule(scheduleId: string): Promise<ScheduleActionResponse> {
  const response = await apiClient.post(`/workflow/schedules/${scheduleId}/resume`)
  return response as unknown as ScheduleActionResponse
}

/**
 * 立即觸發排程
 */
export async function triggerSchedule(scheduleId: string): Promise<TriggerResponse> {
  const response = await apiClient.post(`/workflow/schedules/${scheduleId}/trigger`)
  return response as unknown as TriggerResponse
}

/**
 * 預覽下次執行時間
 */
export async function previewNextRuns(
  scheduleId: string,
  count: number = 5
): Promise<NextRunsPreview> {
  const response = await apiClient.get(
    `/workflow/schedules/${scheduleId}/preview?count=${count}`
  )
  return response as unknown as NextRunsPreview
}

/**
 * 列出執行記錄
 */
export async function listExecutions(params?: {
  schedule_id?: string
  status?: ExecutionStatus
  limit?: number
}): Promise<ExecutionListResponse> {
  const searchParams = new URLSearchParams()
  if (params?.schedule_id) searchParams.set('schedule_id', params.schedule_id)
  if (params?.status) searchParams.set('status', params.status)
  if (params?.limit) searchParams.set('limit', params.limit.toString())

  const query = searchParams.toString()
  const response = await apiClient.get(`/workflow/executions${query ? `?${query}` : ''}`)
  return response as unknown as ExecutionListResponse
}

/**
 * 獲取執行記錄詳情
 */
export async function getExecution(executionId: string): Promise<ReportExecution> {
  const response = await apiClient.get(`/workflow/executions/${executionId}`)
  return response as unknown as ReportExecution
}

// ==================== Utility Functions ====================

/**
 * 格式化頻率顯示
 */
export function formatFrequency(frequency: ScheduleFrequency): string {
  const labels: Record<ScheduleFrequency, string> = {
    daily: '每日',
    weekly: '每週',
    monthly: '每月',
    custom: '自定義',
  }
  return labels[frequency] || frequency
}

/**
 * 格式化狀態顯示
 */
export function formatScheduleStatus(status: ScheduleStatus): string {
  const labels: Record<ScheduleStatus, string> = {
    active: '運行中',
    paused: '已暫停',
    completed: '已完成',
    failed: '失敗',
  }
  return labels[status] || status
}

/**
 * 格式化報告類型顯示
 */
export function formatReportType(reportType: ReportType): string {
  const labels: Record<ReportType, string> = {
    price_analysis: '價格分析',
    competitor_report: '競品報告',
    sales_summary: '銷售摘要',
    inventory_alert: '庫存警報',
    custom: '自定義',
  }
  return labels[reportType] || reportType
}

/**
 * 獲取狀態顏色
 */
export function getStatusColor(status: ScheduleStatus): string {
  const colors: Record<ScheduleStatus, string> = {
    active: 'text-green-600 bg-green-100',
    paused: 'text-yellow-600 bg-yellow-100',
    completed: 'text-blue-600 bg-blue-100',
    failed: 'text-red-600 bg-red-100',
  }
  return colors[status] || 'text-gray-600 bg-gray-100'
}

File: backend/app/schemas/workflow.py
# =============================================
# 工作流 Schema
# =============================================

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID
from pydantic import BaseModel, Field


# =============================================
# 枚舉值常量
# =============================================

class ScheduleFrequency:
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    CUSTOM = "custom"


class ScheduleStatus:
    ACTIVE = "active"
    PAUSED = "paused"
    COMPLETED = "completed"
    FAILED = "failed"


class ExecutionStatus:
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    SKIPPED = "skipped"


class ReportType:
    PRICE_ANALYSIS = "price_analysis"
    COMPETITOR_REPORT = "competitor_report"
    SALES_SUMMARY = "sales_summary"
    INVENTORY_ALERT = "inventory_alert"
    CUSTOM = "custom"


# =============================================
# ScheduledReport Schemas
# =============================================

class ScheduleBase(BaseModel):
    """排程基礎 Schema"""
    name: str = Field(..., min_length=1, max_length=200, description="排程名稱")
    description: Optional[str] = Field(None, max_length=1000, description="排程描述")
    report_type: str = Field(default=ReportType.PRICE_ANALYSIS, description="報告類型")
    report_config: Optional[Dict[str, Any]] = Field(None, description="報告配置")
    frequency: str = Field(default=ScheduleFrequency.DAILY, description="頻率")
    schedule_time: Optional[str] = Field("09:00", pattern=r"^\d{2}:\d{2}$", description="執行時間 HH:MM")
    schedule_day: Optional[int] = Field(None, ge=1, le=31, description="執行日期")
    cron_expression: Optional[str] = Field(None, max_length=100, description="Cron 表達式")
    timezone: str = Field(default="Asia/Hong_Kong", description="時區")
    delivery_channels: Optional[Dict[str, Any]] = Field(None, description="交付渠道配置")


class ScheduleCreate(ScheduleBase):
    """創建排程請求"""
    pass


class ScheduleUpdate(BaseModel):
    """更新排程請求"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)
    report_type: Optional[str] = None
    report_config: Optional[Dict[str, Any]] = None
    frequency: Optional[str] = None
    schedule_time: Optional[str] = Field(None, pattern=r"^\d{2}:\d{2}$")
    schedule_day: Optional[int] = Field(None, ge=1, le=31)
    cron_expression: Optional[str] = Field(None, max_length=100)
    timezone: Optional[str] = None
    delivery_channels: Optional[Dict[str, Any]] = None
    status: Optional[str] = None


class ScheduleResponse(ScheduleBase):
    """排程響應"""
    id: UUID
    status: str
    last_run_at: Optional[datetime] = None
    next_run_at: Optional[datetime] = None
    run_count: int = 0
    success_count: int = 0
    failure_count: int = 0
    consecutive_failures: int = 0
    source_conversation_id: Optional[str] = None
    created_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ScheduleListResponse(BaseModel):
    """排程列表響應"""
    items: List[ScheduleResponse]
    total: int
    limit: int
    offset: int


# =============================================
# ReportExecution Schemas
# =============================================

class ExecutionResponse(BaseModel):
    """執行記錄響應"""
    id: UUID
    schedule_id: UUID
    status: str
    scheduled_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_ms: Optional[int] = None
    report_content: Optional[str] = None
    report_data: Optional[Dict[str, Any]] = None
    delivery_status: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None
    retry_count: int = 0
    created_at: datetime

    class Config:
        from_attributes = True


class ExecutionListResponse(BaseModel):
    """執行記錄列表響應"""
    items: List[ExecutionResponse]
    total: int


# =============================================
# 操作響應
# =============================================

class ScheduleActionResponse(BaseModel):
    """排程操作響應"""
    success: bool
    message: str
    schedule: Optional[ScheduleResponse] = None


class TriggerResponse(BaseModel):
    """觸發執行響應"""
    success: bool
    message: str
    execution_id: Optional[UUID] = None
    task_id: Optional[str] = None


# =============================================
# 預覽
# =============================================

class NextRunsPreview(BaseModel):
    """下次執行時間預覽"""
    schedule_id: UUID
    next_runs: List[datetime]


# =============================================
# AlertWorkflowConfig Schemas
# =============================================

class AlertConfigBase(BaseModel):
    """告警工作流配置基礎 Schema"""
    name: str = Field(..., min_length=1, max_length=200, description="配置名稱")
    is_active: bool = Field(default=True, description="是否啟用")
    trigger_conditions: Optional[Dict[str, Any]] = Field(
        None,
        description="觸發條件 {price_drop_threshold: 10, categories: [...]}"
    )
    auto_analyze: bool = Field(default=True, description="自動執行 AI 分析")
    auto_create_proposal: bool = Field(default=False, description="自動創建改價提案")
    notify_channels: Optional[Dict[str, Any]] = Field(
        None,
        description="通知渠道配置 {telegram: {enabled: true}, email: {...}}"
    )
    quiet_hours_start: Optional[str] = Field(
        None,
        pattern=r"^\d{2}:\d{2}$",
        description="靜默開始時間 HH:MM"
    )
    quiet_hours_end: Optional[str] = Field(
        None,
        pattern=r"^\d{2}:\d{2}$",
        description="靜默結束時間 HH:MM"
    )


class AlertConfigCreate(AlertConfigBase):
    """創建告警配置請求"""
    pass


class AlertConfigUpdate(BaseModel):
    """更新告警配置請求"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    is_active: Optional[bool] = None
    trigger_conditions: Optional[Dict[str, Any]] = None
    auto_analyze: Optional[bool] = None
    auto_create_proposal: Optional[bool] = None
    notify_channels: Optional[Dict[str, Any]] = None
    quiet_hours_start: Optional[str] = Field(None, pattern=r"^\d{2}:\d{2}$")
    quiet_hours_end: Optional[str] = Field(None, pattern=r"^\d{2}:\d{2}$")


class AlertConfigResponse(AlertConfigBase):
    """告警配置響應"""
    id: UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AlertConfigListResponse(BaseModel):
    """告警配置列表響應"""
    items: List[AlertConfigResponse]
    total: int
